add docs for EndpointSlice conditions (#25229)
kubernetes/website: content/en/docs/concepts/services-networking/endpoint-slices.md

* IPv6
* FQDN (Fully Qualified Domain Name)

### Conditions

The EndpointSlice API stores conditions about endpoints that may be useful for consumers.
The three conditions are `ready`, `serving`, and `terminating`.

#### Ready

> `Ready` is a condition that maps to a pod's `Ready` condition. A running pod with the `Ready` condition set to `True`

Reviewer: Nit: Elsewhere in this doc, Kubernetes resource names like "pod" are capitalized as "Pod". I'm not actually sure if that's the standard practice though, so if not, feel free to disregard. Similarly, `True` is capitalized here but is not elsewhere. Not sure what the correct value is, but should be consistent.

Reviewer: One more tiny nit - the rest of this file is being wrapped to 80 char line lengths, definitely not required but would be cool if we could keep that pattern up.

Reviewer 2: After checking the source, I don't think we have condition named `Ready`. The condition name is `ready`, `serving` and `terminating`.

We use `true` for boolean values rather than `True`.

Author: good points, I keep forgetting to use serialized json fields in docs :facepalm:

should have this EndpointSlice condition also set to `true`. For compatibility reasons, `ready` is NEVER `true` when a
pod is terminating. Consumers should refer to the `serving` condition to inspect pod readiness for terminating pods.
The only exception to this rule is for Services with `spec.publishNotReadyAddresses` set to `true`. Endpoints for these

> Services will always have `ready` condition.

Reviewer: ```suggestion
Services will always have the `ready` condition set to `true`.
```

#### Serving

{{< feature-state for_k8s_version="v1.20" state="alpha" >}}

> `Serving` is identical to the `Ready` condition, except it does not account for terminating states. Consumers of the

Reviewer: Does this include the adjustment for `spec.publishNotReadyAddresses`?

Author: Interestingly, I don't think `serving` is always true for `spec.publishNotReadyAddresses`. Do we need to change that for beta?

I see value in keeping `serving` distinct from `publishNotReadyAddresses` but I need to catch up on all the context behind this field.

Reviewer: This feels like a really tricky one. The [comment for `publishNotReadyAddresses`](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3689-L3691) says:

```
The Kubernetes controllers that generate Endpoints and EndpointSlice resources for Services interpret this to mean that all endpoints are considered "ready" even if the Pods themselves are not.
```

As it reads right now, I think it would make sense for the behavior of `serving` to mirror that of `ready` here. It looks like @JeremyOT made the last update to this in https://github.com/kubernetes/kubernetes/pull/90783, any thoughts here? Also interested in perspective from @thockin on how we should treat `serving` here.

Reviewer: If that is the behavior we want, is there any way we could still get this into 1.20 as a bug fix?

Reviewer 3: I think it would be great to know if pods were in fact ready/serving regardless of pNRA.

I don't think the current pNRA behavior is intuitive which is why I updated the docs to explain what actually happens. publishNotReadyAddresses should really be named makeEndpointsAlwaysReady (there's no not-ready endpoints to publish, they're all ready).

It seems like a win to have "serving" always provide an accurate view of the endpoint - especially if that's what it's already doing.

Reviewer: That makes sense, it is a super confusing field. Maybe we'll need another update to the docs around that field once `Serving` becomes available.

Author: Thanks folks, fwiw I feel that `serving` actually removes the need for `spec.publishNotReadyAddreses` altogether. I think that field was mostly added due to old behavior with Endpoints only including ready endpoints and excluding anything else.

Reviewer 3: Isn't pNRA focused on DNS? Endpoints include everything, but DNS still only publishes ready endpoints by default. In an ideal world where we didn't have legacy behavior to account for I think pNRA would just be a signal to the DNS implementation and not affect endpoint readiness at all.

Author: Good point

EndpointSlice API should check this condition if they care about pod readiness while the pod is also terminating.

{{< note >}}

Although `serving` is almost identical to `ready`, it was added to not break the existing meaning of `ready`.
It may be unexpected for existing clients if `ready` could be `true` for terminating endpoints, since historically
terminating endpoints were never included in the Endpoints or EndpointSlice API to begin with. For this reason,

> `ready` is ALWAYS `false` for terminating endpoints, and a new condition `serving` was added in v1.20 so that clients

Reviewer: Is this true when `spec.publishNotReadyAddresses` is true?

can track readiness for terminating pods independent of the existing semantics for `ready`.

{{< /note >}}

#### Terminating

{{< feature-state for_k8s_version="v1.20" state="alpha" >}}

> `Terminating` is a condition that indicates whether an endpoint is terminating. For pods, this is any pod that has a deletion
> timestamp and is slated for terminating by Kubernetes.

Reviewer: Nit: I'm not sure we need to say "and is slated for terminating by Kubernetes". It could be confused as meaning something else must be true in addition to a deletion timestamp being set.
```suggestion
`Terminating` is a condition that indicates whether an endpoint is terminating. For pods, this is any pod that has a deletion
timestamp set.
```

Author: ack, will update

* IPv6
* FQDN (Fully Qualified Domain Name)

### Conditions

The EndpointSlice API stores conditions about endpoints that may be useful for consumers.
The three conditions are `ready`, `serving`, and `terminating`.

#### Ready

`ready` is a condition that maps to a Pod's `Ready` condition. A running Pod with the `Ready`

> condition set to `True` should have this EndpointSlice condition also set to `true`. For

Reviewer 4: @andrewsykim Are the capitalizations of `Ready` and `True` here correct? Based on line 108 ("`ready` condition"), it's not immediately clear to me.

Author: yeah it's a little bit confusing because we are talking about the conditions of two separate APIs. `ready` condition in EndpointSlice is lower-case because it is a json/yaml field. `Ready` condition for Pods is upper-case because it is a codified value for a pod condition.

compatibility reasons, `ready` is NEVER `true` when a Pod is terminating. Consumers should refer
to the `serving` condition to inspect the readiness of terminating Pods. The only exception to
this rule is for Services with `spec.publishNotReadyAddresses` set to `true`. Endpoints for these
Services will always have the `ready` condition set to `true`.

#### Serving

{{< feature-state for_k8s_version="v1.20" state="alpha" >}}

`serving` is identical to the `ready` condition, except it does not account for terminating states.
Consumers of the EndpointSlice API should check this condition if they care about pod readiness while
the pod is also terminating.

{{< note >}}

> Although `serving` is almost identical to `ready`, it was added to not break the existing meaning

Reviewer 4: grammar nit
```suggestion
Although `serving` is almost identical to `ready`, it was added to prevent breaking the existing meaning
```

Author: thanks, will update

of `ready`. It may be unexpected for existing clients if `ready` could be `true` for terminating
endpoints, since historically terminating endpoints were never included in the Endpoints or

> EndpointSlice API to begin with. For this reason, `ready` is ALWAYS `false` for terminating

Reviewer 4: Formatting nit
```suggestion
EndpointSlice API to begin with. For this reason, `ready` is _always_ `false` for terminating
```

Author: thanks, will udpate
